internal/metrics/synthetics/token.go

```diff
@@ -0,0 +1,140 @@
+package synthetics
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/superfly/fly-go"
+	"github.com/superfly/flyctl/gql"
+	"github.com/superfly/flyctl/internal/config"
+	"github.com/superfly/flyctl/internal/flyutil"
+	"github.com/superfly/flyctl/internal/state"
+	"github.com/superfly/flyctl/terminal"
+	"github.com/superfly/macaroon"
+	"github.com/superfly/macaroon/flyio"
+	"github.com/superfly/macaroon/resset"
+)
+
+func generateSyntheticsToken(ctx context.Context) (token string, err error) {
+	client := flyutil.ClientFromContext(ctx)
+	var orgs []fly.Organization
+	if orgs, err = client.GetOrganizations(ctx); err != nil {
+		return
+	}
+
+	var orgsTokens [][]byte
+
+	for _, org := range orgs {
+		var orgToken [][]byte
+
+		orgToken, err = getOrgReadOnlyToken(ctx, org)
+		if err != nil {
+			return
+		}
+
+		orgsTokens = append(orgsTokens, orgToken...)
+	}
+
+	authHeader := macaroon.ToAuthorizationHeader(orgsTokens...)
+
+	return authHeader, nil
+}
+
+func GetSyntheticsToken(ctx context.Context) (token string, err error) {
+	// Prevent synthetics panics from bubbling up to the user.
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic: %v", r)
+		}
+	}()
+
+	cfg := config.FromContext(ctx)
+	if cfg.SyntheticsToken != "" {
+		terminal.Debugf("Config has synthetics token\n")
+		return cfg.SyntheticsToken, nil
+	}
+
+	if cfg.SyntheticsToken == "" && cfg.Tokens.GraphQL() != "" {
+		terminal.Debugf("Generating synthetics token\n")
+		token, err := generateSyntheticsToken(ctx)
+		if err != nil {
+			return "", err
+		}
+		if err = persistSyntheticsToken(ctx, token); err != nil {
+			return "", err
+		}
+		cfg.SyntheticsToken = token
+		return token, nil
+	}
+	return "", errors.New("failed to get synthetics token")
+}
+
+func getOrgReadOnlyToken(ctx context.Context, org fly.Organization) ([][]byte, error) {
+	var (
+		token     string
+		apiClient = flyutil.ClientFromContext(ctx)
+		expiry    = "168h" // 1w
+		profile   = "deploy_organization"
```

Comment:
Since web doesn't know how flyctl is going to attenuate these tokens, it'll think that they're super powerful. They'll show up in `fly tokens list` as such and it might be confusing for users. I'd recommend we make a new token ["profile"](https://github.com/superfly/web/blob/master/app/models/tokens/profiles.rb) in web, so they can be super locked down before flyctl even gets them. I can help with that change if you want.

Reply:
This is great feedback, thanks! I'd like to take a look at the profiles first, I'll ping you if I got stuck. Sounds good?

Reply:
Please let me know if I got it [right](https://github.com/superfly/web/pull/3393)